Crash when a video element is moved to a new document after calling `webkitShowPlaybackTargetPicker`

eric-carlson opened the pull request:
#### bf4c3839062027f9a448cd7f4d70a1858dc2266d
<pre>
Crash when a video element is moved to a new document after calling `webkitShowPlaybackTargetPicker`
<a href="https://bugs.webkit.org/show_bug.cgi?id=296826">https://bugs.webkit.org/show_bug.cgi?id=296826</a>
<a href="https://rdar.apple.com/157029496">rdar://157029496</a>

Reviewed by Jer Noble.

The Document HashMaps that map between MediaPlaybackTargetClient and PlaybackTargetClientContextIdentifier
held raw pointers to MediaPlaybackTargetClient. Update to hold WeakPtrs instead.

* LayoutTests/media/adopt-node-after-showing-picker-crash-expected.txt: Added.
* LayoutTests/media/adopt-node-after-showing-picker-crash.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new test.
* LayoutTests/platform/gtk/TestExpectations: Ditto.
* LayoutTests/platform/wpe/TestExpectations: Ditto.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::willBeRemovedFromFrame):
(WebCore::Document::addPlaybackTargetPickerClient):
(WebCore::Document::removePlaybackTargetPickerClient):
(WebCore::Document::showPlaybackTargetPicker):
(WebCore::Document::playbackTargetPickerClientStateDidChange):
(WebCore::Document::playbackTargetAvailabilityDidChange):
(WebCore::Document::setPlaybackTarget):
(WebCore::Document::setShouldPlayToPlaybackTarget):
(WebCore::Document::playbackTargetPickerWasDismissed):
* Source/WebCore/dom/Document.h:
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.h:
* Source/WebCore/platform/graphics/MediaPlaybackTargetClient.h:

Canonical link: <a href="https://commits.webkit.org/298166@main">https://commits.webkit.org/298166@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a32e3b6c7806319ed5082939ffe1d9bbcbb86af3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114465 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24674 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120631 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65175 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fd2262e3-1f0a-4e2e-8bfd-67aafda841b4) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116354 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34840 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42771 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86987 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41884 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4cb27a3c-e37a-43c1-b3fc-35cf1c02dc7e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117413 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27755 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102796 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67380 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26942 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20923 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64311 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97132 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21037 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123836 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41478 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30949 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95806 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41855 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98987 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95591 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24362 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40753 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18592 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37503 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41357 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46867 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40945 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44252 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42696 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9968,7 +9977,8 @@ void Document::playbackTargetPickerWasDismissed(PlaybackTargetClientContextIdent
 
-    it->value->playbackTargetPickerWasDismissed();
+    if (RefPtr client = it->value.get())
+        client->playbackTargetPickerWasDismissed();
```
Can we also update the map in `HTMLMediaElement::didMoveToNewDocument`?
I think you need to call `removePlaybackTargetPickerClient` on the old document
and call `addPlaybackTargetPickerClient` on the new document.

eric-carlson replied:
I missed this comment when I added the PR to the the merge queue, so I will do what you suggest in a followup

eric-carlson committed:
Crash when a video element is moved to a new document after calling `webkitShowPlaybackTargetPicker`
https://bugs.webkit.org/show_bug.cgi?id=296826
rdar://157029496

Reviewed by Jer Noble.

The Document HashMaps that map between MediaPlaybackTargetClient and PlaybackTargetClientContextIdentifier
held raw pointers to MediaPlaybackTargetClient. Update to hold WeakPtrs instead.

* LayoutTests/media/adopt-node-after-showing-picker-crash-expected.txt: Added.
* LayoutTests/media/adopt-node-after-showing-picker-crash.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new test.
* LayoutTests/platform/gtk/TestExpectations: Ditto.
* LayoutTests/platform/wpe/TestExpectations: Ditto.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::willBeRemovedFromFrame):
(WebCore::Document::addPlaybackTargetPickerClient):
(WebCore::Document::removePlaybackTargetPickerClient):
(WebCore::Document::showPlaybackTargetPicker):
(WebCore::Document::playbackTargetPickerClientStateDidChange):
(WebCore::Document::playbackTargetAvailabilityDidChange):
(WebCore::Document::setPlaybackTarget):
(WebCore::Document::setShouldPlayToPlaybackTarget):
(WebCore::Document::playbackTargetPickerWasDismissed):
* Source/WebCore/dom/Document.h:
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.h:
* Source/WebCore/platform/graphics/MediaPlaybackTargetClient.h:

Canonical link: https://commits.webkit.org/298166@main
